File: rust/plugins/analytics/android/src/main/java/app/nativeblade/analytics/AnalyticsPlugin.kt
package app.nativeblade.analytics

import android.app.Activity
import android.os.Bundle
import app.tauri.annotation.Command
import app.tauri.annotation.TauriPlugin
import app.tauri.plugin.Invoke
import app.tauri.plugin.Plugin
import com.google.firebase.analytics.FirebaseAnalytics

@TauriPlugin
class AnalyticsPlugin(private val activity: Activity) : Plugin(activity) {

    private val analytics by lazy { FirebaseAnalytics.getInstance(activity) }

    @Command
    fun apply(invoke: Invoke) {
        val args = invoke.getArgs()
        val ops = args.optJSONArray("ops")
        if (ops == null) {
            invoke.resolve()
            return
        }

        for (i in 0 until ops.length()) {
            val op = ops.optJSONObject(i) ?: continue
            when (op.optString("op")) {
                "event" -> analytics.logEvent(op.optString("name"), bundleOf(op.optJSONObject("params")))
                "screen" -> {
                    val bundle = Bundle().apply {
                        putString(FirebaseAnalytics.Param.SCREEN_NAME, op.optString("name"))
                    }
                    analytics.logEvent(FirebaseAnalytics.Event.SCREEN_VIEW, bundle)
                }
                "userId" -> analytics.setUserId(op.optString("value").ifEmpty { null })
                "userProperty" -> analytics.setUserProperty(
                    op.optString("key"),
                    op.optString("value").ifEmpty { null }
                )
                "setEnabled" -> analytics.setAnalyticsCollectionEnabled(op.optBoolean("enabled", true))
            }
        }
        invoke.resolve()
    }

    private fun bundleOf(params: org.json.JSONObject?): Bundle {
        val bundle = Bundle()
        if (params == null) return bundle
        val keys = params.keys()
        while (keys.hasNext()) {
            val key = keys.next()
            when (val value = params.get(key)) {
                is String -> bundle.putString(key, value)
                is Int -> bundle.putLong(key, value.toLong())
                is Long -> bundle.putLong(key, value)
                is Double -> bundle.putDouble(key, value)
                is Boolean -> bundle.putLong(key, if (value) 1L else 0L)
                else -> bundle.putString(key, value.toString())
            }
        }
        return bundle
    }
}

File: rust/plugins/analytics/ios/Sources/AnalyticsPlugin.swift
import FirebaseAnalytics
import SwiftRs
import Tauri
import UIKit

enum ParamValue: Decodable {
    case string(String)
    case int(Int)
    case double(Double)
    case bool(Bool)

    init(from decoder: Decoder) throws {
        let c = try decoder.singleValueContainer()
        if let b = try? c.decode(Bool.self) { self = .bool(b); return }
        if let i = try? c.decode(Int.self) { self = .int(i); return }
        if let d = try? c.decode(Double.self) { self = .double(d); return }
        self = .string((try? c.decode(String.self)) ?? "")
    }

    var asAny: Any {
        switch self {
        case .string(let s): return s
        case .int(let i): return i
        case .double(let d): return d
        case .bool(let b): return b
        }
    }
}

class AnalyticsOp: Decodable {
    var op: String
    var name: String?
    var key: String?
    var value: String?
    var enabled: Bool?
    var params: [String: ParamValue]?
}

class ApplyArgs: Decodable {
    var ops: [AnalyticsOp]
}

class AnalyticsPlugin: Plugin {
    @objc public func apply(_ invoke: Invoke) throws {
        let args = try invoke.parseArgs(ApplyArgs.self)

        for op in args.ops {
            switch op.op {
            case "event":
                var params: [String: Any] = [:]
                op.params?.forEach { params[$0.key] = $0.value.asAny }
                Analytics.logEvent(op.name ?? "", parameters: params.isEmpty ? nil : params)
            case "screen":
                Analytics.logEvent(
                    AnalyticsEventScreenView,
                    parameters: [AnalyticsParameterScreenName: op.name ?? ""]
                )
            case "userId":
                Analytics.setUserID(op.value)
            case "userProperty":
                Analytics.setUserProperty(op.value, forName: op.key ?? "")
            case "setEnabled":
                Analytics.setAnalyticsCollectionEnabled(op.enabled ?? true)
            default:
                break
            }
        }

        invoke.resolve()
    }
}

@_cdecl("init_plugin_nativeblade_analytics")
func initPlugin() -> Plugin {
    return AnalyticsPlugin()
}
